qutebrowser/keyinput/eventfilter.py

```diff
@@ -118,8 +120,92 @@ def eventFilter(self, obj: Optional[QObject], event: Optional[QEvent]) -> bool:
             raise
 
 
+class IMEEventHandler(QObject):
+    """Handles the cursorRectangleChanged event from the inputMethod.
+
+    This provides better input detection
+    """
+
+    _last_seen_rect: Optional[QRectF] = None
+
+    def __init__(self, parent: QObject = None) -> None:
+        super().__init__(parent)
+        self._input_method = QApplication.inputMethod()
+        assert self._input_method is not None
+        self._input_method.cursorRectangleChanged.connect(
+            self.cursor_rectangle_changed
+        )
+        self._last_seen_rect = None
+
+    @pyqtSlot()
+    def cursor_rectangle_changed(self) -> None:
+        """Handler for when rectangle changes.
+
+        Event handler for cursorRectangleChanged
+        """
+        # todo:
+        #   clear last_seen_rect on mode exit so that you can click on
+        #     focused input field and re-enter
+        #   last seen rect per window? tab? tab might work better with
+        #      remembering focus for tabs
+        #   anything to unregister? saw some hangs on crash but might be
+        #      because of being a temp basedir
+        # some input examples here https://www.javatpoint.com/html-form-input-types
+        #  <input type="date">: doesn't report as having an input method enabled,
+        #  although the existing heuristics pick it up
+        # if insert_mode_auto_load is false but there is a blinking cursor on
+        # load clicking the scroll bar will enter insert mode
+        assert self._input_method is not None
+
+        new_rect = self._input_method.cursorRectangle()
+        if self._last_seen_rect and self._last_seen_rect.contains(new_rect):
+            print("contains")
```

Comment:
Those debug prints will need to go, qutebrowser can't just print random things in the middle of its logging. It probably makes sense to convert them to proper debug logging statements?